src/main/rx/nrf24_cx10.c

```diff
@@ -0,0 +1,369 @@
+
+static nrf24_protocol_t cx10Protocol;
+
+typedef enum {
+    STATE_BIND = 0,
+    STATE_ACK,
+    STATE_DATA
+} protocol_state_t;
+
+STATIC_UNIT_TESTED protocol_state_t protocolState;
+
+static uint8_t ackCount;
+#define ACK_TO_SEND_COUNT 8
+
+static uint8_t payloadSize;
+
+#define RX_TX_ADDR_LEN     5
+STATIC_UNIT_TESTED uint8_t rxTxAddr[RX_TX_ADDR_LEN] = {0xcc, 0xcc, 0xcc, 0xcc, 0xcc};
+#ifdef USE_PRECONVERTED_XN297
+STATIC_UNIT_TESTED uint8_t txAddr[RX_TX_ADDR_LEN] = {0x55, 0x0F, 0x71, 0x0C, 0x00}; // converted XN297 address, 0xC710F55 (28 bit)
+STATIC_UNIT_TESTED uint8_t rxAddr[RX_TX_ADDR_LEN] = {0x49, 0x26, 0x87, 0x7d, 0x2f}; // converted XN297 address
+#endif
+#define TX_ID_LEN 4
+STATIC_UNIT_TESTED uint8_t txId[TX_ID_LEN];
+
+STATIC_UNIT_TESTED uint8_t rfChannelIndex = 0;
+#define RF_CHANNEL_COUNT 4
+STATIC_UNIT_TESTED uint8_t rfChannels[RF_CHANNEL_COUNT]; // channels are set using txId from bind packet
+
+static uint32_t timeOfLastHop;
+static uint32_t hopTimeout = 5000; // 5ms
+
+void cx10Nrf24Init(nrf24_protocol_t protocol)
+{
+    cx10Protocol = protocol;
+    protocolState = STATE_BIND;
+    payloadSize = (protocol == NRF24RX_CX10) ? CX10_PROTOCOL_PAYLOAD_SIZE : CX10A_PROTOCOL_PAYLOAD_SIZE;
+
+    NRF24L01_Initialize(0); // sets PWR_UP, no CRC
+
+    NRF24L01_WriteReg(NRF24L01_01_EN_AA, 0);
+    NRF24L01_WriteReg(NRF24L01_02_EN_RXADDR, BV(NRF24L01_02_EN_RXADDR_ERX_P0));  // Enable data pipe 0 only
+    NRF24L01_WriteReg(NRF24L01_03_SETUP_AW, NRF24L01_03_SETUP_AW_5BYTES);   // 5-byte RX/TX address
+    NRF24L01_SetChannel(CX10_RF_BIND_CHANNEL);
+
+    NRF24L01_WriteReg(NRF24L01_06_RF_SETUP, NRF24L01_06_RF_SETUP_RF_DR_1Mbps | NRF24L01_06_RF_SETUP_RF_PWR_n12dbm);
+#ifdef USE_PRECONVERTED_XN297
+    // RX_ADDR for pipes P2 to P5 are left at default values
+    NRF24L01_FlushRx();
+    NRF24L01_WriteRegisterMulti(NRF24L01_10_TX_ADDR, txAddr, RX_TX_ADDR_LEN);
+    NRF24L01_WriteRegisterMulti(NRF24L01_0A_RX_ADDR_P0, rxAddr, RX_TX_ADDR_LEN);
+    XN297_SetRxTxAddr(rxAddr, txAddr, RX_TX_ADDR_LEN);
+
+    NRF24L01_WriteReg(NRF24L01_08_OBSERVE_TX, 0x00);
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, payloadSize + 2); // payload + 2 bytes CRC
+
+    NRF24L01_WriteReg(NRF24L01_1C_DYNPD, 0x00); // Disable dynamic payload length on all pipes
+
+    NRF24L01_SetRxMode(); // enter receive mode to start listening for packets
+#else
+    XN297_SetTXAddr(rxTxAddr, RX_TX_ADDR_LEN); // must do before SetRXAddr, even for RX only
+    XN297_SetRXAddr(rxTxAddr, RX_TX_ADDR_LEN);
+    NRF24L01_FlushRx();
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, payloadSize + 2); // payload + 2 bytes CRC
+    NRF24L01_SetRxMode();
+    NRF24L01_WriteReg(NRF24L01_1C_DYNPD, 0x00); // Disable dynamic payload length on all pipes
+#endif
+}
+
+/*
+ * Returns true if it is a bind packet.
+ */
+STATIC_UNIT_TESTED bool checkBindPacket(const uint8_t *packet)
+{
+    const bool bindPacket = (packet[0] == 0xaa);
+    if (bindPacket) {
+        txId[0] = packet[1];
```

Comment:
I thought about that. But I think the code size is probably about the same and I think the explicit setting reads more clearly.